src/index.rs

```diff
@@ -0,0 +1,125 @@
+/// Construct an Index.
+use crate::prelude::{State, TransitionKey};
+use crate::regex::{get_vocabulary_transition_keys, state_scan_tokens};
+use crate::vocabulary::Vocabulary;
+use std::collections::{HashMap, HashSet};
+
+pub type Result<T, E = crate::Error> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct FSMInfo {
+    pub(crate) initial: State,
+    pub(crate) finals: HashSet<State>,
+    pub(crate) transitions: HashMap<(State, TransitionKey), State>,
+    pub(crate) alphabet_anything_value: TransitionKey,
+    pub(crate) alphabet_symbol_mapping: HashMap<String, TransitionKey>,
+}
+
+impl FSMInfo {
+    pub fn new(
+        initial: State,
+        finals: HashSet<State>,
+        transitions: HashMap<(State, TransitionKey), State>,
+        alphabet_anything_value: TransitionKey,
+        alphabet_symbol_mapping: HashMap<String, TransitionKey>,
+    ) -> Self {
+        Self {
+            initial,
+            finals,
+            transitions,
+            alphabet_anything_value,
+            alphabet_symbol_mapping,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Index {
+    pub(crate) initial: u32,
+    finals: HashSet<u32>,
+    states_to_token_subsets: HashMap<u32, HashMap<u32, u32>>,
+    #[allow(dead_code)]
+    eos_token_id: u32,
+}
+
+impl Index {
+    pub fn new(
+        fsm_info: &FSMInfo,
+        vocabulary: &Vocabulary,
+        eos_token_id: u32,
+        frozen_tokens: HashSet<String>,
+    ) -> Result<Self> {
+        let mut states_to_token_subsets: HashMap<u32, HashMap<u32, u32>> = HashMap::new();
+        let mut seen: HashSet<State> = HashSet::new();
+        let mut next_states: HashSet<State> = HashSet::from([fsm_info.initial]);
+
+        let vocabulary_transition_keys = get_vocabulary_transition_keys(
+            &fsm_info.alphabet_symbol_mapping,
+            fsm_info.alphabet_anything_value,
+            vocabulary,
+            &frozen_tokens,
+        );
+
+        while let Some(start_state) = next_states.iter().cloned().next() {
+            next_states.remove(&start_state);
+
+            let token_ids_end_states = state_scan_tokens(
+                &fsm_info.transitions,
+                fsm_info.initial,
+                &fsm_info.finals,
+                vocabulary,
+                &vocabulary_transition_keys,
+                start_state,
+            );
+
+            for (token_id, end_state) in token_ids_end_states {
+                let inner_map = states_to_token_subsets.entry(start_state).or_default();
+                inner_map.insert(token_id, end_state);
+
+                if !seen.contains(&end_state) {
+                    next_states.insert(end_state);
+                }
+            }
+
+            seen.insert(start_state);
+        }
```

Comment:
There might be a step missing here. In the Python version, the call chain is
[`create_states_mapping`](https://github.com/dottxt-ai/outlines-core/blob/9db992785a358af5b921960beb0a2ba0717347d4/python/outlines_core/fsm/guide.py#L105) 
[`create_states_mapping_from_fsm`](https://github.com/dottxt-ai/outlines-core/blob/9db992785a358af5b921960beb0a2ba0717347d4/python/outlines_core/fsm/guide.py#L142)
[`create_fsm_index_tokenizer`](https://github.com/dottxt-ai/outlines-core/blob/9db992785a358af5b921960beb0a2ba0717347d4/python/outlines_core/fsm/regex.py#L437)
[`create_fsm_index_end_to_end`](https://github.com/dottxt-ai/outlines-core/blob/9db992785a358af5b921960beb0a2ba0717347d4/src/python_bindings/mod.rs#L144).

The above code covers all of `create_fsm_index_end_to_end`.

The check below  (`is_valid`) is from `create_states_mapping_from_fsm`.

However, there is another check in `create_fsm_index_tokenizer` that adds transitions for the EOS token to all final states:
https://github.com/dottxt-ai/outlines-core/blob/9db992785a358af5b921960beb0a2ba0717347d4/python/outlines_core/fsm/regex.py#L478-L484

The comment says: `TODO: Do we really need this anymore?`

Well, do we? Because if we don't add the EOS token transitions here, we need to make sure to allow them somewhere else.